src/views/virtualmachines/tree/hooks/useTreeViewData.ts

```diff
@@ -0,0 +1,58 @@
+import { useMemo } from 'react';
+
+import { VirtualMachineModelGroupVersionKind } from '@kubevirt-ui/kubevirt-api/console';
+import { V1VirtualMachine } from '@kubevirt-ui/kubevirt-api/kubevirt';
+import { useIsAdmin } from '@kubevirt-utils/hooks/useIsAdmin';
+import useProjects from '@kubevirt-utils/hooks/useProjects';
+import { isEmpty } from '@kubevirt-utils/utils/utils';
+import { useK8sWatchResource, useK8sWatchResources } from '@openshift-console/dynamic-plugin-sdk';
+import { OBJECTS_FETCHING_LIMIT } from '@virtualmachines/utils';
+
+type UseTreeViewData = {
+  isAdmin: boolean;
+  loaded: boolean;
+  loadError: any;
+  projectNames: string[];
+  vms: V1VirtualMachine[];
+};
+
+export const useTreeViewData = (): UseTreeViewData => {
+  const isAdmin = useIsAdmin();
+
+  const [projectNames, projectNamesLoaded, projectNamesError] = useProjects();
+
+  const [allVMs, allVMsLoaded] = useK8sWatchResource<V1VirtualMachine[]>({
+    groupVersionKind: VirtualMachineModelGroupVersionKind,
+    isList: true,
+    limit: OBJECTS_FETCHING_LIMIT,
+  });
+
+  // user has limited access, so we can only get vms from allowed namespaces
+  const allowedResources = useK8sWatchResources<{ [key: string]: V1VirtualMachine[] }>(
+    Object.fromEntries(
+      projectNamesLoaded && !isAdmin
+        ? (projectNames || []).map((namespace) => [
+            namespace,
+            {
+              groupVersionKind: VirtualMachineModelGroupVersionKind,
+              isList: true,
+              namespace,
+            },
+          ])
+        : [],
+    ),
+  );
+
+  const memoizedVMs = useMemo(
+    () => (isAdmin ? allVMs : Object.values(allowedResources).flatMap((resource) => resource.data)),
+    [allVMs, allowedResources, isAdmin],
+  );
+
+  return {
+    isAdmin,
+    loaded: projectNamesLoaded && (isAdmin ? allVMsLoaded : !isEmpty(memoizedVMs)),
```

Comment:
i would not use `!isEmpty(memoizedVMs)` as it will keep loading in case of non-admin user with no vms. I think we should use 
`Object.values(allowedResources).every((resource) => resource.loaded)` or something like that

Reply:
The reason I'm doing like this is because a user can see a project in the project select, but he can not list vms on that namespace while there are other namespaces he can list vms in, the way you suggest while not show the tree view if there's one namespace that gets an error.
I can better the condition if I do `Object.values(allowedResources).some((resource) => resource.loaded)` wdyt?

Reply:
Yes i like your solution